File: app/src/main/java/com/summerland/android/game/Explosion.java
package com.summerland.android.game;

import android.graphics.Bitmap;
import android.graphics.Canvas;

/**
 * Explosion class, Created by steve on 06/10/16.
 */
public class Explosion extends GameObject {

    Explosion(Bitmap res, int posX, int posY, int sizeW, int sizeH, int columns, int numFrames) {

        x = posX;
        y = posY;
        width = sizeW;
        height = sizeH;

        spriteSheet = res;

        Bitmap[] image = new Bitmap[numFrames];
        for (int col=0,row=0; col < numFrames; col++) {

            // Calculate grid offsets to load images from spriteSheet.
            if (col!=0 && col%columns == 0) row++;
            int xOffset = (col - (row*columns))*width;
            int yOffset = row*height;
            image[col] = Bitmap.createBitmap(spriteSheet, xOffset, yOffset, width, height);
        }

        // load images to animation handler.
        animation = new Animation();
        animation.setFrames(image);
        animation.setDelay(ANIM_DELAY);
        startTime = System.nanoTime();
    }

    @Override
    public void update() {

        if (!animation.isPlayedOnce()){
            animation.update();
        }
    }

    @Override
    public void draw(Canvas canvas) {
        if (!animation.isPlayedOnce()){
            canvas.drawBitmap(animation.getImage(), x, y, null);
        }
    }
}

File: app/src/main/java/com/summerland/android/game/Explosion.kt
package com.summerland.android.game

import android.graphics.Bitmap
import android.graphics.Canvas

/**
 * Explosion class, Created by steve on 06/10/16.
 */
class Explosion internal constructor(res: Bitmap, posX: Int, posY: Int, sizeW: Int, sizeH: Int, columns: Int, numFrames: Int) : GameObject() {

    init {
        x = posX
        y = posY
        width = sizeW
        height = sizeH

        spriteSheet = res

        val image = arrayOfNulls<Bitmap>(numFrames)
        var col = 0
        var row = 0
        while (col < numFrames) {

            // Calculate grid offsets to load images from spriteSheet.
            if (col != 0 && col % columns == 0) row++
            val xOffset = (col - row * columns) * width
            val yOffset = row * height
            image[col] = Bitmap.createBitmap(spriteSheet!!, xOffset, yOffset, width, height)
            col++
        }

        // load images to animation handler.
        animation = Animation()
        animation!!.setFrames(image)
        animation!!.setDelay(GameObject.ANIM_DELAY.toLong())
        startTime = System.nanoTime()
    }

    override fun update() {
        if (!animation!!.isPlayedOnce) {
            animation!!.update()
        }
    }

    override fun draw(canvas: Canvas) {
        if (!animation!!.isPlayedOnce) {
            canvas.drawBitmap(animation!!.image, x.toFloat(), y.toFloat(), null)
        }
    }
}
